compiler/src/iree/compiler/Codegen/Common/FlattenMemRefs.cpp

```diff
@@ -0,0 +1,445 @@
+//===- DecomposeMemRefs.cpp - Decompose memrefs pass implementation -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements decompose memrefs pass. It is an altered version of the
+// upstream GPU/DecomposeMemRefs.cpp file. It adds a new option to not decompose
+// it into 0-rank memrefs but instead single-ranked memrefs.
+//
+// Question to answer at this point:
+// 1. should we disallow memrefs with non-identity layout? also cases where
+// offset != 0 and stride != 1? if so we should update test cases.
+
+// TODO:
+// 1. update memref.subview.
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "Passes.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_DECOMPOSEMEMREFSPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+static void setInsertionPointToStart(OpBuilder &builder, Value val) {
+  if (auto *parentOp = val.getDefiningOp()) {
+    builder.setInsertionPointAfter(parentOp);
+  } else {
+    builder.setInsertionPointToStart(val.getParentBlock());
+  }
+}
+
+/// This is copied from static function affine::mlir::computeProduct.
+/// TODO: enable this function in AffineOps.h
+/// Return the product of `terms`, creating an `affine.apply` if any of them are
+/// non-constant values. If any of `terms` is `nullptr`, return `nullptr`.
+static OpFoldResult computeProduct(Location loc, OpBuilder &builder,
```

Comment:
There might already be a utility for this upstream? (Not a problem if there isn't)